_posts/2015-10-27-resetting-data-after-each-ui-test.md

```diff
@@ -0,0 +1,139 @@
+---
+layout: post
+title: Resetting the application data after each test with Xcode 7 UI Testing
+permalink: /resetting-application-data-after-each-test-with-xcode7-ui-testing
+tags: ios, swift, xcode, testing, ui-testing
+categories: ios
+author: Tomáš Linhart
+---
+
+Apple introduced in Xcode 7 UI testing that allows developers to easily test their user interface, but if you try it yourself you will soon discover that tests are not executed in isolation. The application starts where it was left off and it can be very problematic if your application has a state that you modify through your tests. Eg. some kind of on-boarding, creating records and so on.
+
+If you have experience with [Appium](http://appium.io/) or other test automation frameworks, you know that before each test is executed all content and settings are cleared. So the first thing you could try is executing `NSTask` with some custom shell command and remove the content and settings, but unfortunately `NSTask` is not available on iOS.
+
+If you look at `XCUIApplication`, you will not discover any way to reset the content and settings.
+
+```swift
+public class XCUIApplication : XCUIElement {
+    public func launch()
+    public func terminate()
+    public var launchArguments: [String]
+    public var launchEnvironment: [String : String]
+}
+```
+
+But as you can see, you can set `launchArguments` which allows you to set custom arguments that can be read when the application starts and based on that, the application can remove its content and settings. So we can easily pass an argument to the `setUp()` method.
+
+```swift
+let application = XCUIApplication()
+application.launchArguments = ["--ResetData"]
+application.launch()
+```
+
+After that we need to add a code before `UIApplicationMain` is executed to reset the data. Unfortunately, there is no main.swift on iOS where you could add this code and you have only an annotation on AppDelegate with `@UIApplicationMain`. Fortunately, you can remove this annotation and create a file main.swift where you place a code that you would have in your Objective-C project.
```

Comment:
main.swift is not a problem, what do you think about **add the following lines**?